File: src/constitution_memorizer/web/static/recall_align.js
/* Shared recall alignment (LCS) for Type / Recite accuracy maps.
 * Keep logic in sync with constitution_memorizer.web.recall_align.
 */
(function (global) {
  function normWord(text) {
    return String(text || "")
      .toLowerCase()
      .replace(/[^a-z0-9]/g, "");
  }

  function tokenize(text) {
    const stripped = String(text || "").trim();
    if (!stripped) {
      return [];
    }
    return stripped.split(/\s+/);
  }

  function alignTokens(sourceWords, spokenWords) {
    const src = sourceWords.slice();
    const spk = spokenWords.slice();
    const srcNorm = src.map(normWord);
    const spkNorm = spk.map(normWord);
    const n = srcNorm.length;
    const m = spkNorm.length;

    const dp = Array.from({ length: n + 1 }, () => Array(m + 1).fill(0));
    for (let i = 1; i <= n; i += 1) {
      for (let j = 1; j <= m; j += 1) {
        if (srcNorm[i - 1] && srcNorm[i - 1] === spkNorm[j - 1]) {
          dp[i][j] = dp[i - 1][j - 1] + 1;
        } else {
          dp[i][j] = Math.max(dp[i - 1][j], dp[i][j - 1]);
        }
      }
    }

    const hitIndices = new Set();
    const spokenHitIndices = new Set();
    let i = n;
    let j = m;
    while (i > 0 && j > 0) {
      if (srcNorm[i - 1] && srcNorm[i - 1] === spkNorm[j - 1]) {
        hitIndices.add(i - 1);
        spokenHitIndices.add(j - 1);
        i -= 1;
        j -= 1;
      } else if (dp[i - 1][j] >= dp[i][j - 1]) {
        i -= 1;
      } else {
        j -= 1;
      }
    }

    const extras = [];
    for (let k = 0; k < m; k += 1) {
      if (!spokenHitIndices.has(k) && spkNorm[k]) {
        extras.push(spk[k]);
      }
    }

    const hits = hitIndices.size;
    const total = src.length;
    const percent = total ? Math.round((100 * hits) / total) : 0;

    return {
      sourceWords: src,
      spokenWords: spk,
      hitIndices,
      spokenHitIndices,
      extras,
      hits,
      total,
      percent,
      statsLabel: hits + " / " + total + " recalled · " + percent + "%",
    };
  }

  function alignText(sourceText, spokenText) {
    return alignTokens(tokenize(sourceText), tokenize(spokenText));
  }

  global.RecallAlign = {
    normWord,
    tokenize,
    alignTokens,
    alignText,
  };
})(typeof window !== "undefined" ? window : globalThis);

File: src/constitution_memorizer/web/recall_align.py
"""Token alignment for recall scoring (Type / Recite accuracy maps).

Matches the client helper in ``static/recall_align.js``: normalize words by
lowercasing and stripping non-alphanumerics, then LCS-align spoken/typed
tokens against Bare Act source tokens.
"""

from __future__ import annotations

import re
from dataclasses import dataclass

_NON_ALNUM = re.compile(r"[^a-z0-9]")


def norm_word(text: str) -> str:
    return _NON_ALNUM.sub("", text.lower())


def tokenize(text: str) -> list[str]:
    stripped = text.strip()
    if not stripped:
        return []
    return stripped.split()


@dataclass(frozen=True)
class AlignResult:
    """Alignment of spoken/typed tokens against Bare Act source words."""

    source_words: list[str]
    spoken_words: list[str]
    hit_indices: frozenset[int]
    spoken_hit_indices: frozenset[int]
    extras: list[str]

    @property
    def hits(self) -> int:
        return len(self.hit_indices)

    @property
    def total(self) -> int:
        return len(self.source_words)

    @property
    def percent(self) -> int:
        if not self.total:
            return 0
        return int(round(100 * self.hits / self.total))

    def stats_label(self) -> str:
        return f"{self.hits} / {self.total} recalled · {self.percent}%"


def align_tokens(source_words: list[str], spoken_words: list[str]) -> AlignResult:
    """Longest-common-subsequence alignment on normalized tokens."""
    src_norm = [norm_word(w) for w in source_words]
    spk_norm = [norm_word(w) for w in spoken_words]
    n = len(src_norm)
    m = len(spk_norm)

    dp = [[0] * (m + 1) for _ in range(n + 1)]
    for i in range(1, n + 1):
        for j in range(1, m + 1):
            if src_norm[i - 1] and src_norm[i - 1] == spk_norm[j - 1]:
                dp[i][j] = dp[i - 1][j - 1] + 1
            else:
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])

    hit_indices: set[int] = set()
    spoken_hit_indices: set[int] = set()
    i, j = n, m
    while i > 0 and j > 0:
        if src_norm[i - 1] and src_norm[i - 1] == spk_norm[j - 1]:
            hit_indices.add(i - 1)
            spoken_hit_indices.add(j - 1)
            i -= 1
            j -= 1
        elif dp[i - 1][j] >= dp[i][j - 1]:
            i -= 1
        else:
            j -= 1

    extras = [
        spoken_words[k]
        for k in range(m)
        if k not in spoken_hit_indices and spk_norm[k]
    ]
    return AlignResult(
        source_words=list(source_words),
        spoken_words=list(spoken_words),
        hit_indices=frozenset(hit_indices),
        spoken_hit_indices=frozenset(spoken_hit_indices),
        extras=extras,
    )


def align_text(source_text: str, spoken_text: str) -> AlignResult:
    return align_tokens(tokenize(source_text), tokenize(spoken_text))
